### PR TITLE
Synchronize deletion of records from Salesforce

### PR DESCRIPTION
**Pull request by ahorner**
When a record is deleted from Salesforce, it is maintained for about 15
days in an archive in their system, which we can query on each run. As
we identify new records which have been deleted, we should drop them 
from our system, mirroring the actual state of the data.

### CODE REVIEW

**Comment by danhodos** [comment]
It seems like all of the additions to `Cleaner` here are independent of the existing code within `Cleaner`, which makes me think we have a case for having two separate objects, each with their own responsibility.

File: lib/restforce/db/client.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
+module Restforce
+
+  module DB
+
+    # Restforce::DB::Client is a thin abstraction on top of the default
+    # Restforce::Data::Client class, which adds support for a handful of useful
```
Comment (danhodos) [inline lib/restforce/db/client.rb]: You can't fool me, there's just one useful addition here! :wink:

**Comment by danhodos** [comment]
> When a record is deleted from Salesforce, it is maintained for about 15 days in an archive in their system, which we can query on each run. 

What happens after this? Some sort of hard delete on SF's end? (This is for my own edification.)

File: lib/restforce/db/client.rb (hunk 1 above)
Comment (ahorner) [inline lib/restforce/db/client.rb]: At the time I wrote the comment, I had also added the `getUpdated` endpoint, but found it less useful than our direct queries, so I removed it.

**Comment by ahorner** [comment]
I considered the same thing, but adding another object with the responsibility of cleaning up obsolete records seemed a little excessive to me? This still falls under the purview of what I would consider the `Cleaner` class's responsibility. I _could_ redefine the code to simply make `deleted_salesforce_ids` another component of `invalid_salesforce_ids`, which would perhaps make the relationship between what you're seeing as disparate actions a bit more obvious.

Maybe I'll do that.

Regarding the fate of older deleted records, they are automatically purged from the "recycle bin". It's possible for an administrator to trigger a purge manually, as well, though I probably wouldn't recommend it if you're hoping for the worker daemon to notice your changes.

**Comment by danhodos** [comment]
> I considered the same thing, but adding another object with the responsibility of cleaning up obsolete records seemed a little excessive to me?

Eh, it feels okay to me! I can definitely see where you're coming from in that they're both similar (they delete database records that shouldn't be there anymore), but the object reads as if it has two responsibilities now (even the description of the class includes a conjunction two join its two responsibilities). 

Breaking into two objects would be ever so slightly longer, but each object could be looked at in isolation and you'd know exactly the one (smaller) thing it did. Plus, if the code for either part grows, it will be cleaner to start with.

**Comment by danhodos** [comment]
> See: https://www.salesforce.com/developer/docs/api/Content/sforce_api_calls_getdeleted.htm

Cool link, looks like a lot of caveats about use. I'm going to assume GSP is nowhere near their delete log limits and that limit-based purges probably happen infrequently; however, do we need to tell them not to manually trigger a purge, lest they break the portal?

**Pull request merged**